pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -98,88 +119,250 @@ func (r *MachineSetSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+		if condErr := r.updateSynchronizedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse, reasonFailedToConvertCAPIMachineSetToMAPI, conversionErr.Error(), nil); condErr != nil {
+			return ctrl.Result{}, utilerrors.NewAggregate([]error{conversionErr, condErr})
+		}
+
+		return ctrl.Result{}, conversionErr
+	}
+
+	for _, warning := range warns {
+		logger.Info("Warning during conversion", "warning", warning)
+		r.Recorder.Event(mapiMachineSet, corev1.EventTypeWarning, "ConversionWarning", warning)
+	}
+
+	newMapiMachineSet.Spec.Template.Labels = util.MergeMaps(mapiMachineSet.Spec.Template.Labels, newMapiMachineSet.Spec.Template.Labels)
+
+	newMapiMachineSet.SetNamespace(mapiMachineSet.GetNamespace())
+	// The conversion does not set a resource version, so we must copy it over
+	newMapiMachineSet.SetResourceVersion(mapiMachineSet.GetResourceVersion())
+
+	if !reflect.DeepEqual(newMapiMachineSet.Spec, mapiMachineSet.Spec) || !objectMetaIsEqual(newMapiMachineSet.ObjectMeta, mapiMachineSet.ObjectMeta) {
+		logger.Info("Updating MAPI machine set")
+
+		if err := r.Update(ctx, newMapiMachineSet); err != nil {
+			logger.Error(err, "Failed to update MAPI machine set")
+
+			updateErr := fmt.Errorf("failed to update MAPI machine set: %w", err)
+
+			if condErr := r.updateSynchronizedConditionWithPatch(
+				ctx, mapiMachineSet, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachineSet, updateErr.Error(), nil); condErr != nil {
+				return ctrl.Result{}, utilerrors.NewAggregate([]error{updateErr, condErr})
+			}
+
+			return ctrl.Result{}, updateErr
+		}
+
+		logger.Info("Successfully updated MAPI machine set")
+	} else {
+		logger.Info("No changes detected in MAPI machine set")
+	}
+
+	return ctrl.Result{}, r.updateSynchronizedConditionWithPatch(ctx, mapiMachineSet, corev1.ConditionTrue,
+		consts.ReasonResourceSynchronized, messageSuccessfullySynchronized, &capiMachineSet.Generation)
+}
+
+// convertCAPIToMAPIMachineSet converts a CAPI MachineSet to a MAPI MachineSet, selecting the correct converter based on the platform.
+func (r *MachineSetSyncReconciler) convertCAPIToMAPIMachineSet(capiMachineSet *capiv1beta1.MachineSet, infraMachineTemplate client.Object, infraCluster client.Object) (*machinev1beta1.MachineSet, []string, error) {
+	var converter capi2mapi.MachineSetAndMachineTemplate
+
+	switch r.Platform {
+	case configv1.AWSPlatformType:
+		awsMachineTemplate, ok := infraMachineTemplate.(*awscapiv1beta1.AWSMachineTemplate)
+		if !ok {
+			return nil, nil, fmt.Errorf("%w, expected AWSMachineTemplate, got %T", errUnexpectedInfraMachineTemplateType, infraMachineTemplate)
+		}
+
+		awsCluster, ok := infraCluster.(*awscapiv1beta1.AWSCluster)
+		if !ok {
+			return nil, nil, fmt.Errorf("%w, expected AWSCluster, got %T", errUnexpectedInfraClusterType, infraCluster)
+		}
+
+		converter = capi2mapi.FromMachineSetAndAWSMachineTemplateAndAWSCluster(
+			capiMachineSet, awsMachineTemplate, awsCluster,
+		)
```

Comment:
Why not just return directly here?
```suggestion
		return capi2mapi.FromMachineSetAndAWSMachineTemplateAndAWSCluster(
			capiMachineSet, awsMachineTemplate, awsCluster,
		).ToMachineSet()
```

Reply:
I think my reasoning here was that we can assign in the switch, and return outside of it at the end. 

Not strongly holding this opinion though, is returning from within the switch better? I guess this way we have fewer `nolint` comments ?